### PR TITLE
Added reset forgotten password

### PR DESCRIPTION
**Pull request by andresg4**
## TASKS
- [X] Created passwords controller.
- [X] Configured routes.

## NOTES
- [X] When front-end starts working on this, changes may have to be made to this.

### CODE REVIEW

File: spec/requests/api/v1/passwords/create_spec.rb [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+    end
+  end
+
+  context 'invalid request' do
+    context 'mail not found' do
+      before do
+        params_reset_password[:email] = 'exa@exa.com'
+        post user_password_path, params: params_reset_password, as: :json
+      end
+      it 'returns status 404 not found' do
```
Comment (Alxzu) [inline spec/requests/api/v1/passwords/create_spec.rb:40]: Leave space here

File: spec/requests/api/v1/passwords/create_spec.rb [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+describe 'POST api/v1/users/password', type: :request do
+  let(:params)                { FactoryBot.attributes_for(:user) }
+  let(:params_reset_password) { { email: params[:email], redirect_url: Faker::Internet.url } }
+  before(:each)               { sign_up }
+
+  context 'valid request' do
+    before do
+      post user_password_path, params: params_reset_password, as: :json
+    end
+    it 'returns status 200 OK' do
```
Comment (Alxzu) [inline spec/requests/api/v1/passwords/create_spec.rb:14]: Leave space here

File: spec/requests/api/v1/passwords/create_spec.rb [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+                                               email: params_reset_password[:email]))
+      end
+    end
+
+    context 'without email' do
+      before do
+        params_reset_password = { redirect_url: 'url' }
+        post user_password_path, params: params_reset_password, as: :json
+      end
+      it 'returns status 401 unauthorized' do
```
Comment (Alxzu) [inline spec/requests/api/v1/passwords/create_spec.rb:56]: Leave space here

File: spec/requests/api/v1/passwords/create_spec.rb [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+        expect(json['errors'][0]).to eq(I18n.t('devise_token_auth.passwords.missing_email'))
+      end
+    end
+
+    context 'without redirect url' do
+      before do
+        params_reset_password = { email: params[:email] }
+        post user_password_path, params: params_reset_password, as: :json
+      end
+      it 'returns status 401 unauthorized' do
```
Comment (Alxzu) [inline spec/requests/api/v1/passwords/create_spec.rb:71]: Leave space here

File: spec/requests/api/v1/passwords/edit_spec.rb [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+  before(:each) { sign_up }
+
+  context 'valid request' do
+    before do
+      user = User.find_by_email(params[:email])
+      params_edit_password[:reset_password_token] = user.send_reset_password_instructions
+      params_edit_password[:redirect_url] = Faker::Internet.url
+      get edit_user_password_path, params: params_edit_password
+    end
+    it 'returns status 200 OK' do
```
Comment (Alxzu) [inline spec/requests/api/v1/passwords/edit_spec.rb:19]: Leave space here

File: spec/requests/api/v1/passwords/edit_spec.rb [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+    end
+  end
+
+  context 'invalid request' do
+    before do
+      params_edit_password[:reset_password_token] = Faker::Internet.password(10, 20)
+      params_edit_password[:redirect_url] = Faker::Internet.url
+      get edit_user_password_path, params: params_edit_password
+    end
+    it 'returns status 404 not found' do
```
Comment (Alxzu) [inline spec/requests/api/v1/passwords/edit_spec.rb:41]: Leave space here

File: spec/requests/api/v1/passwords/update_spec.rb [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+  end
+  before(:each) { sign_up }
+
+  context 'valid request' do
+    before do
+      put user_password_path, params: params_update_password, headers: update_password_headers(
+        User.find_by_email(params[:email])
+      ), as: :json
+    end
+    it 'returns status 200 OK' do
```
Comment (Alxzu) [inline spec/requests/api/v1/passwords/update_spec.rb:24]: Leave space here

File: spec/requests/api/v1/passwords/update_spec.rb [hunk 8]
```diff
@@ -0,0 +1,46 @@
...
+      expect(json['message']).to eq(I18n.t('devise_token_auth.passwords.successfully_updated'))
+      expect(json['data']['email']).to eq(params[:email])
+    end
+  end
+
+  context 'invalid request' do
+    before do
+      put user_password_path, params: params_update_password, headers: headers_update, as: :json
+    end
+    it 'returns status 401 not authorized' do
```
Comment (Alxzu) [inline spec/requests/api/v1/passwords/update_spec.rb:40]: Leave space here

**Review by Alxzu — APPROVED** [review]
Nice! Left you a bunch of comments (actually just one). Good jobs! This seems very tidy!

File: app/controllers/api/v1/passwords_controller.rb [hunk 9]
```diff
@@ -0,0 +1,141 @@
...
+          render_edit_success(client_id, token)
+        else
+          render_edit_error
+        end
+      end
+
+      private
+
+      def resource_config_and_save
+        # ensure that user is confirmed
```
Comment (Alxzu) [inline app/controllers/api/v1/passwords_controller.rb]: Don't need to comment the methods, let the code speak by itself ;)

File: app/controllers/api/v1/passwords_controller.rb [hunk 10]
```diff
@@ -0,0 +1,137 @@
...
+
+      def render_create_error(errors)
+        render json: {
+          success: false,
+          errors: errors
+        }, status: 400
+      end
+
+      def render_edit_error
+        render_error(404, 'Not Found')
```
Comment (jdrosales17) [inline app/controllers/api/v1/passwords_controller.rb]: Use i18n here and update tests that check this error message

File: app/controllers/api/v1/passwords_controller.rb [hunk 11]
```diff
@@ -0,0 +1,137 @@
...
+      end
+
+      def render_update_error_unauthorized
+        render_error(401, 'Unauthorized')
```
Comment (jdrosales17) [inline app/controllers/api/v1/passwords_controller.rb]: Same as above

**Review by jdrosales17 — APPROVED** [review]
Nice work!! Approved with two minor comments to address.

**Pull request merged**